File: android/src/main/kotlin/com/arthurmonteiroo/perimeterx_flutter_plugin/PerimeterxHostApiImpl.kt
package com.arthurmonteiroo.perimeterx_flutter_plugin

import android.util.Log
import com.arthurmonteiroo.perimeterx_flutter_android.messages.PerimeterXHostApi
import com.perimeterx.mobile_sdk.PerimeterX
import com.perimeterx.mobile_sdk.PerimeterXChallengeResult

class PerimeterXHostApiImpl : PerimeterXHostApi {

 override fun getHeaders(): Map<String, String> {
  return  PerimeterX.headersForURLRequest()?.toMap() ?: mapOf()
 }

 override fun handleResponse(
  response: String,
  url: String,
  callback: (Result<String>) -> Unit
 ) {
  val handled = PerimeterX.handleResponse(response, null) {
    challengeResult: PerimeterXChallengeResult ->
   callback.invoke(Result.success(if(challengeResult == PerimeterXChallengeResult.SOLVED) "solved" else "cancelled"))
   null
  }
  if(!handled){
   callback.invoke(Result.success("failed"))
  }
 }
}

File: ios/Classes/PerimeterxHostApiImpl.swift
//
//  PerimeterxHostApiImpl.swift
//  perimeterx_flutter_plugin
//
//  Created by Arthur Monteiro Alves Melo on 03/10/24.
//

import Foundation
import PerimeterX_SDK

class PerimeterXHostApiImpl : PerimeterXHostApi {
    func getHeaders() throws -> [String : String] {
        PerimeterX.headersForURLRequest(forAppId: nil) ?? [:]
    }
    
    func handleResponse(response: String, url: String, completion: @escaping (Result<String, any Error>) -> Void) {
        if let data = response.data(using: .utf8), let url = URL(string:url), let httpURLResponse = HTTPURLResponse(url: url, statusCode: 403, httpVersion: nil, headerFields: nil) {
            let handleResponse = PerimeterX.handleResponse(response: httpURLResponse,data: data) {
            challengeResult in
            completion(.success(challengeResult == .solved ? "solved" : "cancelled"))
            }
            
            if !handleResponse {
                completion(.success("failed"))
            }
        } else {
            completion(.success("failed"))
        }
    }
}
